### PR TITLE
gscan: information on hover.

### LINKED ISSUE #1244: cylc gscan: add failed information
From  issue1164 part two:
" Request from BOM:

 * in cases where the suite has failed task, the name of the failed task (for multiple failed tasks, I suppose it could be either the first failed task or number of the failed tasks)"

### PR DESCRIPTION
**Pull request by oliver-sanders**
Closes #1244 

Added tooltip text for the status indicators in gscan. The 5 most recent tasks are displayed, ordered by the most recent of `submitted_time_string`, `started_time_string` and `finished_time_string`. If the row is displaying tasks at a given cycle point the tooltip will show only tasks at that point.

@hjoliver Please Review
@benfitzpatrick Please Review

### CODE REVIEW

**Comment by hjoliver** [comment]
This looks really good, but the tooltip information can be out-of-sync with the main state summary display, because the former is retrieved on-demand and the latter automatically at intervals.  

So, do we need to do one of:
 * force a main display update when retrieving the tool-tip info, or
 * retrieve full tool-tip information - in case it's required - with each state summary update

The second option is probably the right thing to do, otherwise if the suite state has changed since last update you'll retrieve a tool-tip for a different state than intended.

**Comment by oliver-sanders** [comment]
I'd opt for the latter as well.

**Comment by oliver-sanders** [comment]
@hjoliver Task information is now obtained along with the main update procedure so as to keep everything in sync.

**Comment by hjoliver** [comment]
Looks good now.

**Comment by benfitzpatrick** [comment]
I'd rather this didn't grab the state_summary on every update, but just on demand - what do you think?

**Comment by hjoliver** [comment]
(over-the-desktop) - Ben's worried about the additional network traffic, and the volume of state summary information.  

But ... on-demand will necessarily be inconsistent with the summary display sometimes.  What about:

 1. server-side, a new method to supply only the tooltip information 
 2. or report the tooltip info in the scan output!

I like 2.   Note that state summary information can be denied at "public" access level (once gscan follows the CLI scan functionality, to optionally show other users' suites).

**Comment by oliver-sanders** [comment]
#1887 requires state summary information to be acquired on update, also if the user is skimming across the status indicators then if acquiring on-demand a lot of requests will be generated.

Acquiring data on update seems sensible to me, if so are the network traffic issues still a concern?

**Comment by hjoliver** [comment]
@oliver-sanders - I think info on update, not on-demand, is required for consistency.  But rather than retrieving the whole suite state summary we should extend the `identify()` method in `lib/cylc/network/suite_identifier.py` (i.e. the server side scan interface) to report the minimal amount of information needed for the tooltips. Note that this info (and even the state totals) may not be available from other users' suites (although gscan can't display these yet anyway).

**Comment by oliver-sanders** [comment]
@hjoliver @benfitzpatrick 003cb78e6c6a4ffb25856b29b5a244c7ae8fdc98 acquires tooltip information along with suite.identify().

**Comment by hjoliver** [comment]
One suggestion: if the summary ends with "And 1 more" we should just replace that string with the actual result since it takes up the same amount of screen space.

**Comment by oliver-sanders** [comment]
@hjoliver Good point, have implemented.

**Comment by oliver-sanders** [comment]
@hjoliver 3ed3ae733678c47502c04d9c76cf177f391db042 Changes `suite_state` to only send the most recent 6 tasks (hard-coded) for each state.

File: lib/cylc/gui/gscan.py [hunk 1]
```diff
@@ -981,6 +1035,10 @@ def update(self):
             )
             title = suite_info.get("title")
 
+            if 'tasks-by-state' in suite_info:
+                self.tasks_by_state[suite + host] = suite_info[
```
Comment (benfitzpatrick) [inline lib/cylc/gui/gscan.py]: Probably better to use (suite, host) as the key for `tasks_by_state`, as elsewhere.

**Comment by benfitzpatrick** [comment]
Looks good! One comment above, and can we have the numbers of tasks in each state text back in?

**Comment by oliver-sanders** [comment]
@benfitzpatrick The task state summary text is still there, hover over the point string to see it, if the suite is collapsed then the point string is only a few pixels wide and it is a little hard to hover over.

I could change the tooltip text to:

> **`state` (`tasks_in_state`)**
> `task`.`pointstring`
> ...

or:

> _Tasks: `tasks_in_state_1` `state_1`, `tasks_in_state_2` `state_2`, ..._
> **Recently `state` tasks**
> `task`.`pointstring`
> ... 

to better incorporate this information.

**Comment by benfitzpatrick** [comment]
Sorry for the delay - yes, could we have the second one? That looks good.

**Comment by oliver-sanders** [comment]
@benfitzpatrick Changed tooltip text to the second proposed form in 53b6caa40c069393783df765896280696ba48576.

**Comment by hjoliver** [comment]
I checked forward and backward compatibility, and nothing untoward happens, but for new gscan + old suite, the tooltip says "could not get info, try refreshing the scanner".  Is it actually possible for the info to not be supplied until a refresh, or can we assume it must be an older suite daemon and say something like "could not get info; suite running at older cylc version?"

**Comment by oliver-sanders** [comment]
@hjoliver I don't think it is possible for the info to be missing, I just put the exception in for safetys sake. Have updated the text to "could not get info; suite running at older cylc version?".

**Comment by hjoliver** [comment]
Thanks, all good now from my perspective.

**Comment by benfitzpatrick** [comment]
Yep, looks good.

**Comment by benfitzpatrick** [comment]
Just bunging it through the test battery for safety's sake...

**Comment by benfitzpatrick** [comment]
I think the failures are the ones that Matt fixed in #1917.

**Pull request merged**